### ISSUE TITLE
Ongoing PHP errors in /lib/FilesHooks.php during files upload

### ISSUE THREAD

**Issue by MichaIng** [nextcloud/activity#163]
Reported by several users/client-configurations on [help.nextcloud.com](https://help.nextcloud.com/t/nc12-beta2-activity-error-php-undefined-offset-3-at-path-to-nextcloud-apps-activity-lib-fileshooks-php-620/12461?u=michaing).

### Steps to reproduce
1. Update to Nextcloud 12, any version since Beta 2
2. Upload files using desktop (>=2.3.0rc1) or android client (1.4.3 at least)

### Expected behaviour
No error messages should be shown in nextcloud log.

### Actual behaviour
Several identical error messages each minute show up in nextcloud log during upload process (single large AND many smaller files) while no actual misbehaviour was observed so far:
`Error	PHP	Undefined offset: 3 at /path/to/nextcloud/apps/activity/lib/FilesHooks.php#620`

### Server configuration
**Operating system**: Raspbian Jessie

**Web server:** Apache/2.4.25

**Database:** MariaDB 10.0.30

**PHP version:** mod_php 7.0.18-1

**Nextcloud version:** Nextcloud 12 Beta 2 until current 12.0.0.29 release

**Where did you install Nextcloud from:** download.nextcloud.com + web updater

**Signing status:** everything passed integrity check

**List of activated apps:**
```
  - activity: 2.5.2
  - calendar: 1.5.3
  - contacts: 1.5.3
  - dav: 1.3.0
  - federatedfilesharing: 1.2.0
  - files: 1.7.2
  - files_downloadactivity: 1.1.1
  - files_sharing: 1.4.0
  - files_trashbin: 1.2.0
  - files_versions: 1.5.0
  - gallery: 17.0.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.0.0
  - nextcloud_announcements: 1.1
  - notifications: 2.0.0
  - oauth2: 1.0.5
  - ownbackup: 17.5.0
  - previewgenerator: 1.0.6
  - provisioning_api: 1.2.0
  - serverinfo: 1.2.0
  - tasks: 0.9.5
  - twofactor_backupcodes: 1.1.1
  - updatenotification: 1.2.0
  - workflowengine: 1.2.0
```

**Nextcloud configuration:**
```
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "my.domain.org"
        ],
        "datadirectory": "\/mnt\/sda\/ncdata",
        "overwrite.cli.url": "https:\/\/my.domain.org\/nextcloud",
        "dbtype": "mysql",
        "version": "12.0.0.29",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "filelocking.enabled": true,
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "\/var\/run\/redis\/redis.sock",
            "port": 0,
            "dbindex": 0,
            "password": "***REMOVED SENSITIVE VALUE***",
            "timeout": 1.5
        },
        "dbname": "nextcloud",
        "dbhost": "localhost",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "...",
        "loglevel": 2,
        "trashbin_retention_obligation": "disabled",
        "versions_retention_obligation": "disabled"
    }
}
```

### Client configuration
**Operating system:** Windows 10 1073
- Error also shown on Android client

#### Nextcloud log (data/nextcloud.log)
```
Error	PHP	Undefined offset: 3 at /path/to/nextcloud/apps/activity/lib/FilesHooks.php#620
```

**Comment by nickvergessen** [comment]
Fix is in https://github.com/nextcloud/activity/pull/157

**Comment by MichaIng** [comment]
@nickvergessen 
Fix worked great for me and never saw this error again. But another user reported this now on NC 12.0.1 and he checked that the fix (line 141) is applied:
https://help.nextcloud.com/t/nc12-beta2-activity-error-php-undefined-offset-3-at-path-to-nextcloud-apps-activity-lib-fileshooks-php-620/12461/15

Probably during WebDAV + Windows + Excel file change or collabora online.

Any idea if there is maybe another invalid $path variable that needs to be excluded? Maybe via WebDAV (without nextcloud client) excluding ms office temporary files does not work fine or something?

**Comment by danger89** [comment]
Same problem.. Still with Nextcloud 12.0.3rc2.

**Comment by nickvergessen** [comment]
Well the Exception from your log as well as the cron misbehaviour are not caused by this issue here, so no need to panic 😉 

If this still happens for you, please add the following lines before the following code in line 618:
```php
		$path = $node->getPath();
		$sections = explode('/', $path, 4);
		$accessList['ownerPath'] = '/' . $sections[3];
```
before add:
```php
		if (substr_count($node->getPath(), '/') < 3) {
			\OC::$server->getLogger()->logException(new \RuntimeException('Error creating activity for path ' . $path), [
				'app' => 'activity',
			]);
		}
```

Than wait until the error happens again and provide me with that log entry.

**Comment by GitHubUser4234** [comment]
Hi @nickvergessen,

I have applied the patch #157 to NC 12 and can confirm that it doesn't fix the problem when renaming files (seems the error occurs whenever a file is renamed). I have added the suggested debugging lines to the code and this is the output:

`{"reqId":"WktAj6hGTj52wKAGJYqeuAAAAAg","level":3,"time":"2018-01-02T08:19:28+00:00","remoteAddr":"xxx.xxx.xxx.xxx","user":"testuser1","app":"activity","method":"MOVE","url":"\/server\/remote.php\/webdav\/testfile1.txt","message":"Exception: {\"Exception\":\"RuntimeException\",\"Message\":\"Error creating activity for path \\\/\",\"Code\":0,\"Trace\":\"#0 \\\/www\\\/nextcloud\\\/apps\\\/activity\\\/lib\\\/FilesHooks.php(367): OCA\\\\Activity\\\\FilesHooks->getUserPathsFromPath('\\\/', 'testuser1')\\n#1 \\\/www\\\/nextcloud\\\/apps\\\/activity\\\/lib\\\/FilesHooks.php(337): OCA\\\\Activity\\\\FilesHooks->fileRenaming('\\\/testfile1.txt', '\\\/testfile2.txt')\\n#2 \\\/www\\\/nextcloud\\\/apps\\\/activity\\\/lib\\\/FilesHooksStatic.php(79): OCA\\\\Activity\\\\FilesHooks->fileMovePost('\\\/testfile1.txt', '\\\/testfile2.txt')\\n#3 \\\/www\\\/nextcloud\\\/lib\\\/private\\\/legacy\\\/hook.php(106): OCA\\\\Activity\\\\FilesHooksStatic::fileMovePost(Array)\\n#4 \\\/www\\\/nextcloud\\\/lib\\\/private\\\/Files\\\/View.php(829): OC_Hook::emit('OC_Filesystem', 'post_rename', Array)\\n#5 \\\/www\\\/nextcloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Node.php(137): OC\\\\Files\\\\View->rename('\\\/testfile1.txt', '\\\/testfile2.txt')\\n#6 \\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(151): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Node->setName('testfile2.txt')\\n#7 \\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(642): Sabre\\\\DAV\\\\Tree->move('testfile1.txt', 'testfile2.txt')\\n#8 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpMove(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#9 \\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#10 \\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:MOVE', Array)\\n#11 \\\/www\\\/nextcloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#12 \\\/www\\\/nextcloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(71): Sabre\\\\DAV\\\\Server->exec()\\n#13 \\\/www\\\/nextcloud\\\/remote.php(162): require_once('\\\/www\\\/nextcloud\\\/...')\\n#14 {main}\",\"File\":\"\\\/www\\\/nextcloud\\\/apps\\\/activity\\\/lib\\\/FilesHooks.php\",\"Line\":618}","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.0.29"}
{"reqId":"WktAj6hGTj52wKAGJYqeuAAAAAg","level":3,"time":"2018-01-02T08:19:28+00:00","remoteAddr":"xxx.xxx.xxx.xxx","user":"testuser1","app":"PHP","method":"MOVE","url":"\/server\/remote.php\/webdav\/testfile1.txt","message":"Undefined offset: 3 at \/www\/nextcloud\/apps\/activity\/lib\/FilesHooks.php#624","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.0.29"}
{"reqId":"WktAj6hGTj52wKAGJYqeuAAAAAg","level":1,"time":"2018-01-02T08:19:28+00:00","remoteAddr":"xxx.xxx.xxx.xxx","user":"testuser1","app":"admin_audit","method":"MOVE","url":"\/server\/remote.php\/webdav\/testfile1.txt","message":"File renamed: \"\/testfile1.txt\" to \"\/testfile2.txt\"","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.0.29"}
`

 Thanks!

**Comment by nickvergessen** [comment]
Fixed with #226
